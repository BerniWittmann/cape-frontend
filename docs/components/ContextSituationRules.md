# ContextSituationRules

A Component for modelling the Rules of a Context Situation

## Props

<!-- @vuese:ContextSituationRules:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|value|The Context Situation Rules String|`String`|`false`|''|

<!-- @vuese:ContextSituationRules:props:end -->


## Events

<!-- @vuese:ContextSituationRules:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|change|Fired when the rule changes|The new rule string|

<!-- @vuese:ContextSituationRules:events:end -->


## Methods

<!-- @vuese:ContextSituationRules:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|reset|Resets the Rules to the Rule String|-|

<!-- @vuese:ContextSituationRules:methods:end -->


