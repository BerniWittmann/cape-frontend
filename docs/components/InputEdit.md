# InputEdit

A component to edit a text but hidden behind a text with an edit button

## Props

<!-- @vuese:InputEdit:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|value|The Value of the input field|`String`|`true`|-|
|rules|The Rules array that the input field should be validated against|`Array`|`false`||
|size|Size of the controls|`'medium'` / `'small'` / `'mini'`|`false`|'normal'|
|placeholder|The Placeholder for the input field|`String`|`false`|undefined|
|propName|The name the field should be referenced with within the form|`String`|`false`|'text'|

<!-- @vuese:InputEdit:props:end -->


## Events

<!-- @vuese:InputEdit:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|change|Fired when the text changes|The new text|

<!-- @vuese:InputEdit:events:end -->


## Methods

<!-- @vuese:InputEdit:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|showInput|shows the Input field|-|

<!-- @vuese:InputEdit:methods:end -->


