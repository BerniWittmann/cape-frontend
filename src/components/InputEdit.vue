<template>
  <div class="input-edit">
    <span v-if="!isEditing">{{ data.text }}</span>
    <span v-else>
      <el-form ref="textForm" :model="data" label-position="top" :rules="validationRules" status-icon inline :size="size"
               @submit.native.prevent="hideInput" class="form">
        <el-form-item :prop="propName">
          <el-input
                  v-model="data.text"
                  ref="inputField"
                  :size="size"
                  :placeholder="placeholder"
                  @blur="hideInput"
                  @keyup.enter.native="hideInput"
                  @submit.native.prevent="hideInput"
                  class="input"
          ></el-input>
        </el-form-item>
      </el-form>
    </span>
    <el-button
            type="text"
            icon="el-icon-edit"
            @click="showInput"
            v-if="!isEditing"
            class="black-color tag-space"
    ></el-button>
  </div>
</template>

<script>
/* ============
 * Input Edit Component
 * ============
 *
 * A component to edit aa text but hidden behind a text with an edit button
 */

export default {
  model: {
    prop: 'value',
    event: 'change'
  },

  props: {
    value: {
      type: String,
      required: true
    },
    rules: {
      type: Array,
      default: () => []
    },
    size: {
      type: String,
      default: 'normal'
    },
    placeholder: {
      type: String,
      default: undefined
    },
    propName: {
      type: String,
      default: 'text'
    }
  },

  data() {
    return {
      data: {
        text: undefined
      },
      isEditing: false
    }
  },

  computed: {
    validationRules() {
      return {
        [this.propName]: this.rules
      }
    }
  },

  watch: {
    value: {
      deep: true,
      immediate: true,
      handler: function (newValue) {
        this.data.text = newValue
      }
    }
  },

  methods: {
    showInput() {
      this.isEditing = true
      this.$nextTick(() => {
        try {
          this.$refs.inputField.$refs.input.focus()
        } catch (e) {}
      })
    },

    hideInput() {
      this.$refs.textForm.validate((valid) => {
        if (valid) {
          this.isEditing = false
        }
      })
    }
  }
}

</script>

<style scoped lang="scss">
.input-edit {
  display: inline;
}
.form {
  margin-top: 10px;
  display: inline;
}
.black-color {
  color: #000000;
}

.input {
  width: 300px;
}

.tag-space {
  margin-left: 5px;
}
</style>
