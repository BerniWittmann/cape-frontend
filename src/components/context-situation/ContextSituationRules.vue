<template>
  <input-edit class="context-situation-rules"
          :value="contextSituation.rules" size="mini" :rules="validationRules"
          :placeholder="$t('context_situation.edit.rules')"
  ></input-edit>
</template>

<script>
/* ============
 * ContextSituationRules
 * ============
 *
 *
 */
import InputEdit from '@/components/InputEdit.vue'

export default {
  components: {
    InputEdit
  },

  props: {
    contextSituation: {
      type: Object,
      required: true
    }
  },

  data() {
    return {
      validationRules: [
        { required: true, message: this.$t('context_situation.edit.validation.rules.required'), trigger: 'blur' }
      ]
    }
  }
}
</script>

<style scoped lang="scss">
</style>
