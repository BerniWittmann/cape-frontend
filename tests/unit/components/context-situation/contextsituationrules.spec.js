import { mount } from '@vue/test-utils'
import { i18n } from '../../setupPlugins'

import InputEdit from '@/components/InputEdit.vue'
import ContextSituationRules from '@/components/context-situation/ContextSituationRules.vue'

describe('Components', () => {
  describe('ContextSituationRules', () => {
    let propsData
    let cmp
    let router
    let store

    beforeEach(() => {
      propsData = {
        contextSituation: {
          id: '142',
          name: 'My Context Situation',
          rules: 'The Rules String',
          tags: [{
            id: '43',
            name: 'Second Tag',
            color: '#FFFF00'
          }, {
            id: '44',
            name: 'Third Tag',
            color: '#0FFF00'
          }]
        }
      }
      store = {
        state: {
          contextSituation: {
            activeContextSituation: {
              id: '142',
              name: 'My Context Situation',
              rules: 'The Rules String'
            }
          }
        },
        dispatch: jest.fn().mockImplementation(() => ({
          then: (arg) => arg()
        }))
      }
      router = {
        push: jest.fn()
      }
      render()
    })

    function render() {
      cmp = mount(ContextSituationRules, {
        i18n,
        propsData,
        mocks: {
          $store: store,
          $router: router
        }
      })
    }

    describe('context situation rules management', () => {
      beforeEach(() => {
        store.state.contextSituation.activeContextSituation = undefined
      })

      it('renders', () => {
        expect(cmp.html()).toMatchSnapshot()
      })

      it('shows the rules of the context situation', () => {
        expect(cmp.text()).toContain('The Rules String')
      })

      describe('allows to edit the rules of the context situation', () => {
        it('renders the input edit component for the name', () => {
          const input = cmp.find(InputEdit)
          expect(input.exists()).toBeTruthy()
        })
        it('open input for the rules', (done) => {
          const editRulesButton = cmp.find('button.black-color')
          expect(editRulesButton.exists).toBeTruthy()
          editRulesButton.trigger('click')
          cmp.vm.$nextTick(() => {
            expect(cmp.html()).toMatchSnapshot()
            expect(cmp.find('.el-input__inner').exists).toBeTruthy()
            done()
          })
        })

        it('hides the input after editing the rules', (done) => {
          const editRulesButton = cmp.find('button.black-color')
          expect(editRulesButton.exists).toBeTruthy()

          editRulesButton.trigger('click')
          cmp.vm.$nextTick(() => {
            expect(cmp.html()).toMatchSnapshot()
            const input = cmp.find('.el-input__inner')
            input.setValue('My new Rule')
            input.trigger('keyup', { key: 'Enter' })
            cmp.vm.$nextTick(() => {
              expect(cmp.html()).toMatchSnapshot()
              done()
            })
          })
        })

        it('does not hid the input after editing the rules if its invalid', (done) => {
          const editRulesButton = cmp.find('button.black-color')
          expect(editRulesButton.exists).toBeTruthy()

          editRulesButton.trigger('click')
          cmp.vm.$nextTick(() => {
            expect(cmp.html()).toMatchSnapshot()
            const input = cmp.find('.el-input__inner')
            input.setValue('My new Rule')
            input.trigger('keyup', { key: 'Enter' })
            cmp.vm.$nextTick(() => {
              expect(cmp.html()).toMatchSnapshot()
              done()
            })
          })
        })

        it('checks the input to not be empty', (done) => {
          const editRulesButton = cmp.find('button.black-color')
          editRulesButton.trigger('click')
          cmp.vm.$nextTick(() => {
            expect(cmp.html()).toMatchSnapshot()
            const input = cmp.find('.el-input__inner')
            input.setValue('')
            input.trigger('keypress', { key: 'Enter' })
            expect(cmp.html()).toMatchSnapshot()
            expect(cmp.find('.el-input__inner').exists).toBeTruthy()
            cmp.vm.$nextTick(() => {
              expect(cmp.html()).toMatchSnapshot()
              done()
            })
          })
        })
      })
    })
  })
})
