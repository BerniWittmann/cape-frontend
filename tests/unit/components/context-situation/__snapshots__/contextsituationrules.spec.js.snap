// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components ContextSituationRules context situation rules management allows to edit the rules of the context situation checks the input to not be empty 1`] = `
<div class="input-edit context-situation-rules"><span><form class="el-form form el-form--label-top el-form--inline"><div class="el-form-item el-form-item--feedback is-required el-form-item--mini"><!----><div class="el-form-item__content"><div class="input el-input el-input--mini"><!----><input type="text" autocomplete="off" placeholder="context_situation.edit.rules" class="el-input__inner"><!----><!----><!----></div><!----></div></div></form></span>
  <!---->
</div>
`;

exports[`Components ContextSituationRules context situation rules management allows to edit the rules of the context situation checks the input to not be empty 2`] = `
<div class="input-edit context-situation-rules"><span><form class="el-form form el-form--label-top el-form--inline"><div class="el-form-item el-form-item--feedback is-required el-form-item--mini"><!----><div class="el-form-item__content"><div class="input el-input el-input--mini"><!----><input type="text" autocomplete="off" placeholder="context_situation.edit.rules" class="el-input__inner"><!----><!----><!----></div><!----></div></div></form></span>
  <!---->
</div>
`;

exports[`Components ContextSituationRules context situation rules management allows to edit the rules of the context situation checks the input to not be empty 3`] = `
<div class="input-edit context-situation-rules"><span><form class="el-form form el-form--label-top el-form--inline"><div class="el-form-item el-form-item--feedback is-required el-form-item--mini"><!----><div class="el-form-item__content"><div class="input el-input el-input--mini"><!----><input type="text" autocomplete="off" placeholder="context_situation.edit.rules" class="el-input__inner"><!----><!----><!----></div><!----></div></div></form></span>
  <!---->
</div>
`;

exports[`Components ContextSituationRules context situation rules management allows to edit the rules of the context situation does not hid the input after editing the rules if its invalid 1`] = `
<div class="input-edit context-situation-rules"><span><form class="el-form form el-form--label-top el-form--inline"><div class="el-form-item el-form-item--feedback is-required el-form-item--mini"><!----><div class="el-form-item__content"><div class="input el-input el-input--mini"><!----><input type="text" autocomplete="off" placeholder="context_situation.edit.rules" class="el-input__inner"><!----><!----><!----></div><!----></div></div></form></span>
  <!---->
</div>
`;

exports[`Components ContextSituationRules context situation rules management allows to edit the rules of the context situation does not hid the input after editing the rules if its invalid 2`] = `
<div class="input-edit context-situation-rules"><span>My new Rule</span> <button type="button" class="el-button black-color tag-space el-button--text">
    <!----><i class="el-icon-edit"></i>
    <!----></button></div>
`;

exports[`Components ContextSituationRules context situation rules management allows to edit the rules of the context situation hides the input after editing the rules 1`] = `
<div class="input-edit context-situation-rules"><span><form class="el-form form el-form--label-top el-form--inline"><div class="el-form-item el-form-item--feedback is-required el-form-item--mini"><!----><div class="el-form-item__content"><div class="input el-input el-input--mini"><!----><input type="text" autocomplete="off" placeholder="context_situation.edit.rules" class="el-input__inner"><!----><!----><!----></div><!----></div></div></form></span>
  <!---->
</div>
`;

exports[`Components ContextSituationRules context situation rules management allows to edit the rules of the context situation hides the input after editing the rules 2`] = `
<div class="input-edit context-situation-rules"><span>My new Rule</span> <button type="button" class="el-button black-color tag-space el-button--text">
    <!----><i class="el-icon-edit"></i>
    <!----></button></div>
`;

exports[`Components ContextSituationRules context situation rules management allows to edit the rules of the context situation open input for the rules 1`] = `
<div class="input-edit context-situation-rules"><span><form class="el-form form el-form--label-top el-form--inline"><div class="el-form-item el-form-item--feedback is-required el-form-item--mini"><!----><div class="el-form-item__content"><div class="input el-input el-input--mini"><!----><input type="text" autocomplete="off" placeholder="context_situation.edit.rules" class="el-input__inner"><!----><!----><!----></div><!----></div></div></form></span>
  <!---->
</div>
`;

exports[`Components ContextSituationRules context situation rules management renders 1`] = `
<div class="input-edit context-situation-rules"><span>The Rules String</span> <button type="button" class="el-button black-color tag-space el-button--text">
    <!----><i class="el-icon-edit"></i>
    <!----></button></div>
`;
